src/System.Private.CoreLib/src/Internal/Runtime/Augments/RuntimeThread.Windows.cs

```diff
@@ -36,6 +42,202 @@ internal IntPtr[] GetWaitedHandleArray(int requiredCapacity)
+            }
+        }
+
+        private bool SetPriority(ThreadPriority priority)
+        {
+            if (_osHandle.IsInvalid)
+            {
+                return true;
+            }
+            return Interop.mincore.SetThreadPriority(_osHandle, (int)MapToOSPriority(priority));
+        }
+
+        /// <summary>
+        /// Checks if the underlying OS thread has finished execution.
+        /// </summary>
+        private bool HasFinishedExecution()
+        {
+            uint result = Interop.mincore.WaitForSingleObject(_osHandle, dwMilliseconds: 0);
+            return result == (uint)Interop.Constants.WaitObject0;
+        }
+
+        private bool JoinCore(int millisecondsTimeout)
+        {
+            SafeWaitHandle waitHandle = _osHandle;
+            int result;
+
+            waitHandle.DangerousAddRef();
+            try
+            {
+                result = WaitHandle.WaitForSingleObject(waitHandle.DangerousGetHandle(), millisecondsTimeout);
+            }
+            finally
+            {
+                waitHandle.DangerousRelease();
+            }
+
+            return result == (int)Interop.Constants.WaitObject0;
+        }
+
+        private void StartCore(object parameter)
+        {
+            using (LockHolder.Hold(_lock))
+            {
+                if (!GetThreadStateBit(ThreadState.Unstarted))
+                {
+                    throw new ThreadStateException(SR.ThreadState_AlreadyStarted);
+                }
+
+                // TODO: OOM hardening, _maxStackSize
+                GCHandle threadHandle = GCHandle.Alloc(this);
+                _threadStartArg = parameter;
+
+                try
+                {
+                    uint threadId;
+
+                    _osHandle = new SafeWaitHandle(Interop.mincore.CreateThread(IntPtr.Zero, IntPtr.Zero,
+                        AddrofIntrinsics.AddrOf<Interop.mincore.ThreadProc>(StartThread), (IntPtr)threadHandle, 0, out threadId),
+                        ownsHandle: true);
+
+                    // Ignore errors (as in CoreCLR)
+                    SetPriority(_priority);
+
+                    // Wait until the new thread is started (as in CoreCLR)
+                    while (GetThreadStateBit(ThreadState.Unstarted))
+                    {
+                        Yield();
+                    }
+                }
+                finally
+                {
+                    if (_osHandle == null)
+                    {
+                        threadHandle.Free();
+                        _threadStartArg = null;
+                    }
+                }
+            }
+        }
+
+        [NativeCallable(CallingConvention = CallingConvention.StdCall)]
+        private static uint StartThread(IntPtr parameter)
+        {
+            GCHandle threadHandle = (GCHandle)parameter;
+            RuntimeThread thread = (RuntimeThread)threadHandle.Target;
+            t_currentThread = thread;
```

Comment:
The CoreCLR contract is that the started thread either clears the Thread::TS_Unstarted flag or sets the Thread::TS_FailStarted flag.  In the former case the started thread is responsible for the cleanup, in the latter case the starting thread does the cleanup.
Since the `t_currentThread = thread` assignment may throw on CoreRT, we need to put it under try/catch.